ipd/0020/README.adoc

```diff
@@ -0,0 +1,770 @@
+separate from the module-under-test. They should use `modlmisc`
+linkage and perform test registration/deregistration in their
+`_init(9E)` and `_fini(9E)` callbacks. A given test module will
+typically live adjacent to its module-under-test in the `usr/src/uts`
+tree. The source file and binary should generally use the name
+`<module-under-test>_test`. You should deviate from this rule when the
+module covers many subsystems and breaking it up would add clarity.
+For example, the mblk routines in the "STREAMS subsystem" are part of
+`genunix`. But `genunix` covers a lot of ground, and `genunix_test.c`
+would be a pretty big source file. It makes more sense to create a
+`stream_test.c` next to the `stream.c` file and create a `stream_test`
+module that execises the various stream APIs in `genunix`.
+
+Test modules, like system libraries, should come welded to the system
+-- the source code for the test module should live in illumos-gate.
+The main exception would be a test delivered as part of an out-of-gate
+driver or for downstream distributions testing their own kernel
+functionality (though in that case it should be in their downstream
+gate).
+
+Delivering test modules is a choice left to each downstream
+distribution. That said, we must make a default decision about how to
+structure the IPS manifests in gate. First, it seems to make sense to
+at least give the ktest facility its own package, which includes only
+the means to register, list, and execute tests, but does not deliver
+any tests itself. Things get more interesting when determining how
+test modules should be delivered. The following is a table of
+potential options and their tradeoffs.
+
+|===
+|Delivery| Tradeoffs
+
+|1. All in-gate tests delivered in ktest package. Deliver all in-gate
+ test modules as part of the ktest package.
+a|* One package gives you everything.
+* No test modules delivered unless you absoltely want them.
+* Delivers test modules for modules that may not be attached and that
+have no relevance to your system .
+
+|2. Each test module is delivered with whatever package delivers the
+ module-under-test. Each package which delivers a test module has a
+ dependency on ktest facility package.
+a|* Only the necessary test
+ modules are installed.
+* Probably makes the most logical sense.
+* Given that at least one module-under-test is part of the main kernel
+  (like genunix), this effectively means ktest is always delivered.
+
+|3. Same as previous, but don't require ktest dependency.
+a|* Same benefits as above, but test execution can only happen if the
+user decides to also install ktest. Otherwise the test modules lay
+doormant on the filesystem (not loaded).
+
+|===
+
+I think we should go with option (3). We should deliver test-modules
+with their module-under-test, but only load/run them when the ktest
+facility is installed (and even then they would not be loaded until
+the user specifically requests that one or more test-modules be
+loaded).
```

Comment:
This definitely sounds like the best approach, and I would also add an IPS facet to these modules so that, for the distributions that use IPS, they can easily by turned off or never installed at all, in the same way that we currently facet header files (and other development-only components) and man pages.
Almost all IPS facets are on by default, we may wish to consider one that is off by default; that might require some minor tweaks to IPS itself.

Reply:
I did not know about facets. This sounds like a great idea. Thanks.

Reply:
Added a blurb about using a `facet.test` tag. Thanks for the idea! I'm not sure how to broach the second part, about having some facets be `false` by default. I'm also not really sure how this all works with building a distro: e.g., could you disable the facet when building your installation image? Would that allow it to be `false` to start out with? What happens if someone wants that data after the fact? If it's in system-kernel can they enable the facet and then someone get that missing content?

Reply:
Most facets default to true/on, but any that start with `facet.debug.` or `facet.optional.` default to off (the trailing . is there to encourage namespace use).

If you'd like them not installed by default, I suggest you use something like `facet.debug.ktest` for these modules (or `.optional.`), or have them installed by default but they're easily removed en-bloc with `pkg change-facet facet.ktest=False`

Reply:
When building a distribution image, you can definitely set facets to specific values between the `pkg image-create` step and the first `pkg install` step, and those facets will never be installed.  I agree that `facet.debug.ktest` seems appropriate for this.